rfcs/20190411-string-unification.md

```diff
@@ -0,0 +1,302 @@
+
+### Eigen Compatibility
+
+Of the four string types defined above, the only type with potential eigen
+compatibility issues is the `OffsetType`.  Since the `OffsetType` relies on an
+offset value instead of a char pointer to point to the character string, and
+since eigen is restrictive on how
+[values are indexed](https://github.com/eigenteam/eigen-git-mirror/blob/branches/3.3/unsupported/Eigen/CXX11/src/Tensor/TensorMap.h#L147),
+the simplest approach for
+providing eigen compatibility is to define the 'offset' value as an offset from
+the this pointer of a tstring scalar, and not as an offset from the start of
+the tensor buffer.  More concretely, accessing the character string for an
+OffsetType would be analogous to:
+
+```cpp
+    const char* data() const {
+      return reinterpret_cast<const char*>(this) + offset_;
+    }
+```
+
+### Lvalue assignment
+
+In the scenario were an assignment exceeds the capacity `OffsetType`, `SmallType`,
+or `PreallocType` the string is converted to `LargeType` string.  The original
+string type is copied as a prefix to the `LargeType` so that it can be reverted
+when the capacity falls below the original on a subsequent assignment.  This
+feature will allow for lvalue assignment of OffsetType types, which, as a
+corollary, will allow for the assignment of TFLite string tensors.
+
+### Comparison of SSO implementations in gcc, llvm, and MSVC
+
+|                                           | Size in Bytes | Small String Capacity
+|------------------------------------------ | --            | --
+|tensorflow::string                         | 16            | 15
+|tensorflow::string w/ capacity (see below) | 24            | 23
+|GCC                                        | 32            | 15
+|MSVC                                       | 32            | 15
+|LLVM                                       | 24            | 22
+
+### Capacity member variable
+
+Currently, TFLite has an ~8 byte overhead per string entry, which is used to
+describe the offset.  The tstring class described above has an overhead of 16
+bytes, and does not include a capacity field normally found in SSO
+implementations for LargeType strings.  This comes with downsides.  Without a
+capacity field LargeType strings are forced to always call realloc on
+assignment.  To reduce the potential number of calls to realloc, we can add a
+capacity field, at the cost of increasing the per string overhead to 24 bytes.
+This would put as at parity with LLVM strings, but would result in a 3x
+overhead compared to current TFLite strings.
+
+### Updates to C API
+
+Currently, downstream language bindings to
+[Java](https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/java/src/main/native/tensor_jni.cc?#L270),
+[golang](https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/go/tensor.go?#L395),
+etc, are expected to
+modify a raw buffer in order to construct and pass a string tensor.  We propose
```

Comment:
Not quite; this is specifically about the C API, and the fact that accessors/mutators do not exist in the C-API for C string tensor construction.

In the Java case, string tensors are created in the bridge JNI code by directly manipulating a buffer (though nothing precludes one from creating it in Java and passing it through via JNI).  With the purposed update, the string tensors will _still_ be created in the JNI bridge code, but instead of directly manipulating a buffer as is currently done [1], the JNI bridge will use new methods in the TF C API for string tensor construction.

[1] https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/java/src/main/native/tensor_jni.cc?#L246

Reply:
Ok. I think it could still be useful though to have a string layout simple enough to let the Java client (and other language bindings) to write them directly into the tensor buffer, using some custom writers. We might achieve better performances by doing this and that is what we will probably do for all other datatypes. What do you think?

Reply:
For the creation of string tensors, of the four string sub-types proposed, the kOffsetType fits the bill of being a "simple" layout which allows for direct writing to a buffer.  Creating an kOffsetType string tensor would mitigate the current expensive conversion of TF_STRING -> DT_STRING.  The layout is similar to the current C-String/TF_STRING layout (i.e. single block of contiguous memory), but has endianness defined and is compatible with eigen.

Things get a bit complicated for reading and modifying existing string tensors---since in those situations, the string could be any of the four types, and memory ownership comes into play with kLargeType.

**Ideally**, in order to future proof potential changes in byte layout, it would still be beneficial to utilize the new proposed functions in the C-API when reading/modifying/constructing string tensors.  We will provide functions for creating string tensors from lists of strings in order to reduce function call overhead.

**Alternatively**, since we no longer will be using std::string as a string container, and the layout of strings will be well-defined and reside in TF,  we _could_ have duplicate implementations in java for better performance.  For full duplication, kLargeTypes may need an additional field for a deallocator.

In short, this design does not preclude your suggestions of direct access to DT_STRING buffers, and, in fact, enables it (since the ABI of std::string is not defined).  We can assess the potential performance gains vs additional complexity in having a duplicate Java implementation once the DT_STRING scalar is mapped to tensorflow::tstring instead of std::string.

Reply:
Perfect, thanks for the detailed answer